_posts/2021-01-30-Writing-Better-Code-Cloud-Code-with-Parse-Server-Example.md

```diff
@@ -0,0 +1,191 @@
+    user.set('object',secondObject);
+  }
+  // object will always be request.object
+});
+```
+
+## Object Literal Shorthand Syntax
+
+With ES6, you can declare variables much clearer using shorthand syntax.
+
+ES5:
+
+```
+Parse.Cloud.beforeSave('test', function(request) {
+  var object = request.object;
+  var user = request.user;
+  request.log.info(object, user);
+});
+```
+
+ES6 shorthand syntax:
+
+```
+Parse.Cloud.beforeSave('test', ({object,user,log}) => {
+  log.info(object, user);
+});
+```
+
+## For of loops
+
+With ES6, you can write loops using `for (const object of objects)` rather than `for (var i=0; i<objects.length; i++)`
+
+## Promise.all
+
+Unless your code requires series excecution (i.e one task after an other), you should try to leverage `Promise.all` in your cloud functions. This will allow your cloud code to resolve much faster.
+
+Poorly Optimized:
+
+```javascript
+Parse.Cloud.define('testFunction', ({user, params:{name, type}}) => {
+  const nameQuery = new Parse.Query('Name');
+  nameQuery.equalTo('name', name);
+  const nameData = await nameQuery.first({useMasterKey:true});
+  if (!nameData) {
+    throw new Parse.Error(141, 'name does not exist');
+  }
+  const typeQuery = new Parse.Query('Type');
+  typeQuery.equalTo('type',type);
+  const typeData = await typeQuery.first({useMasterKey:true});
+  if (!typeData) {
+    throw new Parse.Error(141, 'type does not exist');
+  }
+  user.set('type', typeData);
+  user.set('name', nameData);
+  await user.save(null, {useMasterKey: true});
+  return 'Type and name are valid'
+},{
+  requireUser:true,
+  fields: ['name','type']
+});
+```
+
+This code is poorly optimized as it resolves the query for `name`, and then the query for `type`. These should be resolved in parallel as they are not dependant on eachother.
+
+Using Promise.all:
+```javascript
+Parse.Cloud.define('testFunction', ({user, params:{name, type}}) => {
+  const nameQuery = new Parse.Query('Name');
+  nameQuery.equalTo('name', name);
+  const typeQuery = new Parse.Query('Type');
+  typeQuery.equalTo('type',type);
+  const [nameData, typeData] = await Promise.all([nameQuery.first({useMasterKey:true}),typeQuery.first({useMasterKey:true})]);
+  if (!nameData) {
+    throw new Parse.Error(141, 'name does not exist');
+  }
+  if (!typeData) {
+    throw new Parse.Error(141, 'type does not exist');
+  }
+  user.set('type', type);
+  user.set('name', name);
+  await user.save(null, {useMasterKey: true});
+  return 'Type and name are valid'
+},{
+  requireUser:true,
+  fields: ['name','type']
+});
+```
+
+These are only a few examples of how to write better cloud code. On the latest [Parse Server Example](https://github.com/parse-community/parse-server-example) project, you can now run `npm run prettier` to style your code, followed by `npm run lint` , which will run over your cloud code and check for quality issues. You can also run `npm run lint-fix` to fix any easy errors, such as indentation.
```

Comment:
It would be nice to add a link here to learn more about ES6, I assume there are good free resources out there!